File: src/practice/Robox/Rob03.java
package practice.Robox;

import java.util.*;

public class Rob03 {


    public static int largestSubgrid(List<List<Integer>> grid, int maxSum) {
        int n = grid.size();

        HashMap<Integer, Integer> maxSums = new HashMap<>();

        for(int size = 1; size < n; size++){
            int step = n - size + 1;
            int maxSizeSum = 0;
            for(int i = 0; i < step; i++){
                for(int j =0; j < step; j++){
                    int curSum = sum(grid, i, j, size);
                    if(curSum > maxSizeSum){
                        maxSizeSum = curSum;
                    }
                }
            }
            maxSums.put(size, maxSizeSum);
        }

        int curMaxSum = -1;
        int curSize = -1;
        for(Integer key: maxSums.keySet()){
            int curSum = maxSums.get(key);
            if(curSum <= maxSum && curSum > curMaxSum){
                curMaxSum = curSum;
                curSize = key;
            }
        }

        return curSize;
    }

    private static int sum(List<List<Integer>> grid, int rS, int cS, int size){
        int n = grid.size();
        int sum = 0;

        for(int r = rS; r < rS+size; r++){
            for( int c= cS; c < cS+size; c++){
                sum += grid.get(r).get(c);
            }
        }
        return sum;
    }


    public static void main(String[] args){
        // List<List<Integer>> grid1 = {{1, 1, 1}, {1, 1, 1}, {1, 1, 1}};


    }
    
}

File: src/practice/Robox/Rob03.py
#%%

def largestSubgrid(grid, maxSum: int) -> int:
    n: int = len(grid)
    sizeWithMaxSum = {}

    for size in range(n):
        step = n - size + 1;
        maxCurSizeSum = 0

        for rStart in range(step):
            for cStart in range(step):
                curSizeSum = sum(grid, rStart, cStart, size)
                if curSizeSum > maxCurSizeSum :
                    maxCurSizeSum = curSizeSum

        sizeWithMaxSum[size] = maxCurSizeSum

    maxMaxSum = -1
    maxSize = -1
    for size, sizeMaxSum in sizeWithMaxSum.items():
        if sizeMaxSum <= maxSum and sizeMaxSum > maxMaxSum :
            maxMaxSum = sizeMaxSum
            maxSize = size

    return maxSize


def sum(grid, rs:int, cs:int, size:int) -> int:
    n = len(grid)
    sum = 0

    for r in range(rs, rs+size):
        for c in range(cs, cs+size):
            sum += grid[r][c]

    return sum


grid1 = [[1,1,1],[1,1,1],[1,1,1]]
print(grid1)
print(largestSubgrid(grid1, 4))

grid2 = [[1,1,1],[2,2,2],[3,3,3]]
print(grid2)
print(largestSubgrid(grid2, 3))
